src/main/scala/com/emc/ecs/spark/sql/sources/s3/relation.scala

```diff
@@ -99,27 +125,48 @@ private class BucketMetadataRDD(
     val projectedAttributes: Seq[String] =
       requiredColumns.map(schema(_))
       .filter(_.metadata.getString(SqlMetadataKeys.MetadataType) == SYSMD.toString)
+      .filterNot(_.name == "Content")
       .map(_.metadata.getString(SqlMetadataKeys.MetadataName))
 
     val request = new QueryObjectsRequest(bucketName)
       .withQuery(query)
       .withAttributes(projectedAttributes)
 
-    paged(request).flatMap(result => result.getObjects.map(convertToRow))
+    implicit val executionContext = ExecutionContext.fromExecutorService(Executors.newFixedThreadPool(10))
+    paged(request).flatMap(p=> readRows(p))
   }
+  case class ObjectData(key: String, usermd: Map[String,String], sysmd: Map[String,String], content: Future[Array[Byte]])
 
-  case class ObjectData(key: String, usermd: Map[String,String], sysmd: Map[String,String])
 
   /**
-    * Convert the given object metadata to a Spark SQL row object.
+    * Convert the given query result to Spark Sql rows.
+    *
+    * @param result a page of query results.
+    * @return a sequence of rows
     */
-  private def convertToRow(obj: QueryObject): Row = {
-    val data = ObjectData(
-      obj.getObjectName,
-      obj.getQueryMds.find(_.getType == USERMD).map(_.getMdMap.toMap).getOrElse[Map[String,String]](Map.empty),
-      obj.getQueryMds.find(_.getType == SYSMD).map(_.getMdMap.toMap).getOrElse[Map[String,String]](Map.empty))
+  private def readRows(result: QueryObjectsResult)(implicit execctx: ExecutionContext): Seq[Row] = {
+    val objects = result.getObjects.map { obj =>
+      ObjectData(
+        obj.getObjectName,
+        obj.getQueryMds.find(_.getType == USERMD).map(_.getMdMap.toMap).getOrElse[Map[String,String]](Map.empty),
+        obj.getQueryMds.find(_.getType == SYSMD).map(_.getMdMap.toMap).getOrElse[Map[String,String]](Map.empty),
+        if(withContent) download(obj.getObjectName) else null)
```

Comment:
Please change `withContent` to `requiredColumns.contains("Content")` - otherwise we're downloading the content even when the Content field isn't selected.